File: samples/HelloWorld/quicklua/QSprite.lua
--[[/*
 * (C) 2012-2013 Marmalade.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */--]]

--[[
Notes on connection between Sprite and Texture:

QSprite
    cocos2d::CCNode* (CCSprite2*, public cocos2d::CCSprite*)
    QAnimation* m_pAnimation;
    QAnimate* m_pAnimateAction; // used to animate sprite if it has > 1 frame
    cocos2d::CCRepeatForever* m_pRunAction; // used to encapsulate the AnimateAction

QAnimate (public cocos2d::CCAnimate)
    // Animation controller, simply a wrapper around CCAnimate

QAnimation
    cocos2d::CCAnimation* m_Animation;
    // Animation, simply a wrapper for CCAnimation
    // CCAnimation contains CCSpriteFrames
    // CCSpriteFrame contains CCTexture2D

To get current CCTexture for sprite:

int f = QAnimate::GetFrame()                // gets current frame ID
CCSpriteFrame* pSF = QAnimation::GetFrame(f)    // gets CCSpriteFrame from fram ID
CCTexture2D* pTex = pSF->getTexture()       // gets CCTexture2D

We really need to store QAtlas* as userdata on the CCTexture2D
USE CCObject:m_uID
--]]

--------------------------------------------------------------------------------
-- Sprite
-- NOTE: This file must have no dependencies on the ones loaded after it by
-- openquick_init.lua. For example, it must have no dependencies on QDirector.lua
--------------------------------------------------------------------------------
QSprite = {}
table.setValuesFromTable(QSprite, QNode) -- previous class in hierarchy
QSprite.__index = QSprite

--------------------------------------------------------------------------------
-- For overriding specific keys to do more complex function calls
-- Currently just used for width and height assignment to scale sprites
-- Falls back to QNode version then to userdata/default behaviour
-- These are applied in initSprite()

function QSprite_set(t, name, value)
    if name == "xScale" then
        t._xScale = value --store user value to avoid rounding errors
        QNode_set(t, "xScale", t._relativeScaleX*value)
    elseif name == "yScale" then
        t._yScale = value
        QNode_set(t, "yScale", t._relativeScaleY*value)
    elseif name == "w" then
        local oldScale = t.xScale
        t._relativeScaleX = value/t._fileW
        QNode_set(t, "xScale",  oldScale * t._relativeScaleX)
        QNode_set(t, name, value)
    elseif name == "h" then
        local oldScale = t.yScale
        t._relativeScaleY = value/t._fileH
        QNode_set(t, "yScale",  oldScale * t._relativeScaleY)
        QNode_set(t, name, value)
    else
        QNode_set(t, name, value)
    end
end

--Using same style as QNode already does for similar overrides
local oldNodeMTI = getmetatable(quick.QNode).__index

function QSprite_get(t, name)
    if name == "xScale" then
        return t._xScale --could return oldNodeMTI(t, name) / t._relativeScaleX, but may get rounding errors
    elseif name == "yScale" then
        return t._yScale
    else
        return oldNodeMTI(t, name)
    end
end

--------------------------------------------------------------------------------
-- Private API
--------------------------------------------------------------------------------
QSprite.serialize = function(o)
	local obj = serializeTLMT(getmetatable(o), o)
    table.setValuesFromTable(obj, serializeTLMT(getmetatable(quick.QSprite), o))
	return obj
end

--[[
/*
Initialise the peer table for the C++ class QSprite.
This must be called immediately after the QSprite() constructor.
*/
--]]
function QSprite:initSprite(n)
    --override metamethods for node (see above)
    getmetatable(n).__newindex = QSprite_set
    getmetatable(n).__index = QSprite_get
	local np = {}
    local ep = tolua.getpeer(n)
    table.setValuesFromTable(np, ep)
	setmetatable(np, QSprite)
	tolua.setpeer(n, np)

    local mt = getmetatable(n) 
    mt.__serialize = QSprite.serialize
end

--------------------------------------------------------------------------------
-- Public API
--------------------------------------------------------------------------------
--[[
/**
Create a sprite node, specifying arbitrary input values.
@param values Lua table specifying name/value pairs.
@return The created sprite.
*/
--]]
function director:createSprite(values)
end
--[[
/**
Create a sprite node, at the specified screen position, with the specified texture name.
@param x X coordinate of sprite.
@param y Y coordinate of sprite.
@param source Either a full filename (including extension) of the texture or an animation to associate with
@return The created sprite.
*/
--]]
function director:createSprite(x, y, source)
    local n = quick.QSprite()
    QNode:initNode(n)
    QSprite:initSprite(n)
    self:addNodeToLists(n)

    local frame = nil
    local w = nil
    local h = nil

    if type(x) == "table" then
        -- Copy any "source" input to local
        source = x["source"]
        x["source"] = nil
        
        --avoid setting w & h (via overrides) until sprite is set-up
        if x.w then
            w = x.w
            x.w = nil
        end
        if x.h then
            h = x.h
            x.h = nil
        end
        
        table.setValuesFromTable(n, x)
        if x.frame then
            frame = x.frame
        end
    else
        dbg.assertFuncVarTypes({"number", "number", "string"}, x, y, source)
        n.x = x
        n.y = y
    end
    n.source = source

    -- See if we need to create a real source
    if type(source) == "string" then
        -- "source" is assumed to point to a single image file
        -- Create an atlas and animation from this file
        local atlas = director:createAtlas(source)
        n.source = director:createAnimation( { start = 1, count = 1, atlas = atlas } )
        n.animationlock = n.source
    end

    -- Allow the texture (Atlas) to be unspecified at the time of creation
    --dbg.assert(n.source, "Source data is nil")
    if n.source then
        n.animation = n.source
        n.source = nil
    end
    if frame then
        n.frame = frame
    end
    
    -- Set width and height after image has been loaded
    n._relativeScaleX = 1
    n._relativeScaleY = 1
    n._fileW = n.w
    n._fileH = n.h
    n.xScale = n._xScale or 1.0
    n.yScale = n._yScale or 1.0
    
    --We have overriden the assignment (_index) operator,
    --so assigning it now will update internals
    if w then
        n.w = w
    end
    if h then
        n.h = h
    end

    return n
end

-- PRIVATE
function director:_createSpriteNoCCNode()
    local n = quick.QSprite(false)
    QNode:initNode(n)
    QSprite:initSprite(n)
    self:addNodeToLists(n)
	return n
end

--[[
/*! Play the current assigned animation.
    @param n (optional) A table containing playback parameters
        startFrame (optional) = The first frame of the animation to play
        loopCount (optional) = The number of times to play the animation. 0 = infinate
*/
--]]
function QSprite:play(n)
	dbg.assertFuncVarTypes({"nil", "table"}, n)

    local startFrame = 1
    local loopCount = 0

    if type(n) == "table" then
        if n.startFrame ~= nil then
            startFrame = n.startFrame
        end
        if n.loopCount ~= nil then
            loopCount = n.loopCount
        end
    end

    self:_play( startFrame, loopCount)
end

--[[
/*! Set the width and height to the file size of the texture.
    xScale and yScale will not be reset.
*/
--]]
function QSprite:resetSize()
    self.w = self._fileW
    self.h = self._fileH
end

    

File: source/QSprite.cpp
/*
 * (C) 2012-2013 Marmalade.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

//#include "AppDelegate.h"
#include "QDirector.h"
#include "QScene.h"
#include "QSceneTextureBatch.h"
#include "QSprite.h"
#include "QLuaHelpers.h"

#include "cocos2d.h"

USING_NS_CC;

// From CCSprite.cpp
#if CC_SPRITEBATCHNODE_RENDER_SUBPIXEL
#define RENDER_IN_SUBPIXEL
#else
#define RENDER_IN_SUBPIXEL(__ARGS__) (ceil(__ARGS__))
#endif

QUICK_NAMESPACE_BEGIN;

//------------------------------------------------------------------------------
// CCSprite2
//------------------------------------------------------------------------------
void CCSprite2::drawMain()
{
    QScene* pScene = g_QDirector->m_DrawingScene->m_QScene;
    QAssert(pScene, "No QScene object");
    if (pScene->m_UseBatching == false)
    {
        CCSprite::draw();
        return;
    }

    // Batching code

    CCTexture2D* pCCTex = getTexture();
    QSceneTextureBatch* pSTB = pScene->GetTextureBatch(pCCTex);
    if (!pSTB)
    {
        pSTB = pScene->AddTextureBatch(pCCTex);
        pSTB->m_FirstSprite = this; // TODO, THINK ABOUT THIS IN TERMS OF SPRITE LIFECYCLE?
    }

    // The CCSprite's texture must point to the texture of the batch's atlas
    m_pobTextureAtlas = pSTB;
    m_pobTexture = pSTB->getTexture();

    // Code dup'd from CCSprite::updateTransform
    // This writes the vertex coords into the atlas's array
    
    // Calculate transform to WORLD (was transform to BATCH)
    //m_transformToBatch = nodeToParentTransform();
    CCAffineTransform xform = nodeToWorldTransform();

    CCSize size = m_obRect.size;

    float x1 = m_obOffsetPosition.x;
    float y1 = m_obOffsetPosition.y;

    float x2 = x1 + size.width;
    float y2 = y1 + size.height;
    float x = xform.tx;
    float y = xform.ty;

    float cr = xform.a;
    float sr = xform.b;
    float cr2 = xform.d;
    float sr2 = -xform.c;
    float ax = x1 * cr - y1 * sr2 + x;
    float ay = x1 * sr + y1 * cr2 + y;

    float bx = x2 * cr - y1 * sr2 + x;
    float by = x2 * sr + y1 * cr2 + y;

    float cx = x2 * cr - y2 * sr2 + x;
    float cy = x2 * sr + y2 * cr2 + y;

    float dx = x1 * cr - y2 * sr2 + x;
    float dy = x1 * sr + y2 * cr2 + y;

    m_sQuad.bl.vertices = vertex3( RENDER_IN_SUBPIXEL(ax), RENDER_IN_SUBPIXEL(ay), m_fVertexZ );
    m_sQuad.br.vertices = vertex3( RENDER_IN_SUBPIXEL(bx), RENDER_IN_SUBPIXEL(by), m_fVertexZ );
    m_sQuad.tl.vertices = vertex3( RENDER_IN_SUBPIXEL(dx), RENDER_IN_SUBPIXEL(dy), m_fVertexZ );
    m_sQuad.tr.vertices = vertex3( RENDER_IN_SUBPIXEL(cx), RENDER_IN_SUBPIXEL(cy), m_fVertexZ );

    // Write vertices to atlas array
    m_uAtlasIndex = pSTB->m_AtlasIndex++;
    m_pobTextureAtlas->updateQuad(&m_sQuad, m_uAtlasIndex);
}
//------------------------------------------------------------------------------
void CCSprite2::drawSetup()
{
    CC_NODE_DRAW_SETUP();
//    ccGLBlendFunc(m_blendFunc.src, m_blendFunc.dst);
}
//------------------------------------------------------------------------------
void CCSprite2::draw()
{
    // Set clipping
    float clip_x = 0.0f;
    float clip_y = 0.0f;
    float clip_w = 0.0f;
    float clip_h = 0.0f;

    QNode *node = m_Sprite;
    while ((clip_w == 0.0f || clip_h == 0.0f) && node != NULL)
    {
        clip_x = node->clipX;
        clip_y = node->clipY;
        clip_w = node->clipW;
        clip_h = node->clipH;

        node = node->m_Parent;
    }

    if (clip_w > 0.0f && clip_h > 0.0f)
        QDirector::SetClippingRect( clip_x, clip_y, clip_w, clip_h);

    // Call out main draw function, which is dup'ed from Cocos2dx
    drawMain();

    // Reset clipping
    if (clip_w > 0.0f && clip_h > 0.0f)
        QDirector::ResetClippingRect();
 
    // Main debug draw
    if (m_Sprite->debugDraw == true)
    {
        QVec2 vertices[4]={
            QVec2(m_sQuad.tl.vertices.x,m_sQuad.tl.vertices.y),
            QVec2(m_sQuad.bl.vertices.x,m_sQuad.bl.vertices.y),
            QVec2(m_sQuad.br.vertices.x,m_sQuad.br.vertices.y),
            QVec2(m_sQuad.tr.vertices.x,m_sQuad.tr.vertices.y),
        };
        QDirector::DrawColor4F(m_Sprite->debugDrawColor.r, m_Sprite->debugDrawColor.g, m_Sprite->debugDrawColor.b, 1);
        QDirector::DrawPolyVert2F((const ccVertex2F*)vertices, 4, true);
    }
    if (m_Sprite->debugDrawTextureRegion == true)
    {
        // draw texture box
        CCSize s = this->getTextureRect().size;
        CCPoint offsetPix = this->getOffsetPosition();
        CCPoint vertices[4] = {
            ccp(offsetPix.x,offsetPix.y), ccp(offsetPix.x+s.width,offsetPix.y),
            ccp(offsetPix.x+s.width,offsetPix.y+s.height), ccp(offsetPix.x,offsetPix.y+s.height)
        };
        QDirector::DrawColor4F(1,0.75,0,1); // orange
        QDirector::DrawPolyVert2F((const ccVertex2F*)vertices, 4, true);
    }

    // Physics draw
    if  (
        (m_Sprite->physics) &&
        (physics::g_Sim->debugDraw)
        )
        m_Sprite->physics->draw();
}

//------------------------------------------------------------------------------
// QSprite
//------------------------------------------------------------------------------
QSprite::QSprite(bool createCCNode)
{
    xFlip = false;
    yFlip = false;
    uvRect.x = uvRect.y = 0.0f;
    uvRect.w = uvRect.h = 1.0f;

    blendMode = ""; // empty string means we get blendMode from the associated QAtlas
    timeScale = 1.0f;
    raisesAnimEvents = false;
    debugDrawTextureRegion = false;

	// Cocos stuff
	if (createCCNode == true)
    {
	    m_CCNode = new CCSprite2(this);
	    ((CCSprite*)m_CCNode)->init();
        m_CCNode->setAnchorPoint(ccp(0, 0));
    }
	else
	{
		// By default, if we are not creating a CC node, we
		// flag that we're not managing any CC node
		m_ManageCCNodeLifeCycle = false;
	}
    m_pAnimation = NULL;
    m_pAnimateAction = NULL;
    m_pRunAction = NULL;
    
    m_pFilter = NULL;
}
//------------------------------------------------------------------------------
QSprite::QSprite(CCSprite *sprite)
{
    uvRect.x = uvRect.y = 0.0f;
    uvRect.w = uvRect.h = 1.0f;

    blendMode = ""; // empty string means we get blendMode from the associated QAtlas
    timeScale = 1.0f;
    raisesAnimEvents = false;
    debugDrawTextureRegion = false;

    m_CCNode = new CCSprite2(this);
    ((CCSprite*)m_CCNode)->initWithTexture(sprite->getTexture());
    m_CCNode->setAnchorPoint(ccp(0, 0));

    CCPoint pos = sprite->getPosition();
    x = pos.x;
    y = pos.y;
    CCRect bounds = sprite->boundingBox();
    w = bounds.size.width;
    h = bounds.size.height;
    xScale = sprite->getScaleX();
    yScale = sprite->getScaleY();
    xSkew = sprite->getSkewX();
    ySkew = sprite->getSkewY();
    rotation = sprite->getRotation();
    xFlip = sprite->isFlipX();
    yFlip = sprite->isFlipY();

    m_pAnimation = NULL;
    m_pAnimateAction = NULL;
    m_pRunAction = NULL;

    m_pFilter = NULL;
}
//------------------------------------------------------------------------------
QSprite::~QSprite()
{
    CC_SAFE_DELETE(m_pFilter);
    
    if ( m_pAnimateAction)
    {
        m_CCNode->stopAction(m_pRunAction);
        CC_SAFE_RELEASE_NULL(m_pRunAction);
    }
    CC_SAFE_RELEASE_NULL(m_pAnimateAction);

//    QTrace("QSprite destroyed");
}
//------------------------------------------------------------------------------
CCSpriteFrame* QSprite::GetSpriteFrame() const
{
    // Note: Return values from m_pAnimateAction->GetFrame() are 1-based
    int f = m_pAnimateAction ? (m_pAnimateAction->GetFrame()-1) : 0;
    QAssert(m_pAnimation, "Sprite has no animation");
    CCSpriteFrame* pSF = m_pAnimation->GetFrame(f); // frame input is 0-based
    return pSF;
}
//------------------------------------------------------------------------------
void QSprite::sync()
{
    // Only sync if sync is enabled
    if (!isSynced)
        return;

    QNode::sync();
    
    // TODO - IMPROVE PERFORMANCE... WE DON'T WANT TO SEARCH FOR THE TEXTURE IN THIS LIST EVERY FRAME!
	// Check we've loaded the texture
	if (m_CCNode)
	{
        CCSprite* pCCNode = (CCSprite*)m_CCNode;

        // Update the width and height from the content
        w = pCCNode->getContentSize().width;
        h = pCCNode->getContentSize().height;

        uint8 alpha8 = (uint8)(_alphaInternal * color.a); // multiply "alpha" by color.a component
        if (pCCNode->getOpacity() != alpha8)
            pCCNode->setOpacity( alpha8);

		pCCNode->setColor(*(ccColor3B*)&color); // cast ccColor4B* to ccColor3B* should be OK

		pCCNode->updateTransform();

        if (pCCNode->isFlipX() != xFlip)
            pCCNode->setFlipX(xFlip);

        if (pCCNode->isFlipY() != yFlip)
            pCCNode->setFlipY(yFlip);

        // Process the blend function setup
        cocos2d::ccBlendFunc blend_func;
        GetBlendFunction( &blend_func);
        pCCNode->setBlendFunc( blend_func);
        
        // Handle filter change
        if (m_LastFilter != filter.name)
        {
            CC_SAFE_DELETE(m_pFilter);
            m_pFilter = QFilter::create(&filter, pCCNode);
            m_LastFilter = filter.name;
        }

        if (m_pFilter)
            m_pFilter->sync();

        // UV rect
		if	(
			(uvRect.x != 0.0f) ||
			(uvRect.y != 0.0f) ||
			(uvRect.w != 1.0f) ||
			(uvRect.h != 1.0f)
			)
		{
			// Results will be probably be undesired if
			// sprite has an animation (strictly speaking, if
			// the SpriteFrame doesn't cover whole texture)
			CCSpriteFrame* pSF = GetSpriteFrame();
			CCRect sfr = pSF->getRectInPixels();
#ifdef _DEBUG
			if	(
				(sfr.origin.x != 0.0f) ||
				(sfr.origin.y != 0.0f) ||
                (sfr.size.width != pCCNode->getTexture()->getPixelsWide()) ||
				(sfr.size.height != pCCNode->getTexture()->getPixelsHigh())
				)
			{
                if  (
                    (uvRect.x < 0.0f) || (uvRect.x >= 1.0f) ||
                    (uvRect.y < 0.0f) || (uvRect.y >= 1.0f) ||
                    (uvRect.w < 0.0f) || (uvRect.w >= 1.0f) ||
                    (uvRect.h < 0.0f) || (uvRect.h >= 1.0f)
                    )
                {
    				QWarning("Sprite frame does not cover whole bitmap, and uvRect dimensions not within unit square: may have undesired results.");
                }
			}
#endif
	        CCRect tr(  sfr.origin.x + (uvRect.x * sfr.size.width),
                        sfr.origin.y + (uvRect.y * sfr.size.height),
                        uvRect.w * sfr.size.width,
                        uvRect.h * sfr.size.height);
			pCCNode->setTextureRect(tr);

            // Update Sprite w,h from SpriteFrame and uvRect
            w = sfr.size.width * uvRect.w;
            h = sfr.size.height * uvRect.h;
		}
	}
}
//------------------------------------------------------------------------------
QAtlas* QSprite::getAtlas() const
{
    CCSpriteFrame* pSF = GetSpriteFrame();
    cocos2d::CCTexture2D* pTex = pSF->getTexture();
    QAtlas* pAtlas = (QAtlas*)pTex->m_uID;
    return pAtlas;
}
//------------------------------------------------------------------------------
void QSprite::pause()
{
    if ( m_pAnimateAction != NULL)
    {
        m_pAnimateAction->SetPlaying( false);
    }
}

//------------------------------------------------------------------------------
void QSprite::_play( int startFrame, int loopCount)
{
    if ( m_pAnimateAction != NULL)
    {
        if ( startFrame > 0)
        {
            set_frame( startFrame);
        }
        m_pAnimateAction->SetLoopCount( loopCount);
        m_pAnimateAction->SetPlaying( true);
    }
}
//------------------------------------------------------------------------------
void QSprite::setFrame( int frame)
{
    set_frame( frame);
}

//------------------------------------------------------------------------------
void QSprite::set_isPlaying(bool value)
{
    if ( m_pAnimateAction != NULL)
    {
        m_pAnimateAction->SetPlaying( value);
    }
}
//------------------------------------------------------------------------------
bool QSprite::get_isPlaying(void)
{
    return (m_pAnimateAction != NULL) && (m_pAnimateAction->IsPlaying());
}
//------------------------------------------------------------------------------
void QSprite::setAnimation(QAnimation *anim)
{
    set_animation( anim);
}
//------------------------------------------------------------------------------
void QSprite::set_animation(QAnimation *anim)
{
    // Release the old animator
    if ( m_pAnimateAction)
    {
        m_CCNode->stopAction(m_pRunAction);
        CC_SAFE_RELEASE_NULL(m_pRunAction);
    }
    CC_SAFE_RELEASE_NULL(m_pAnimateAction);

    m_pAnimation = anim;

    CCSprite* pCCNode = (CCSprite*)m_CCNode;
    CCSpriteFrame *pSpriteFrame = m_pAnimation->GetFrame(0);
	pCCNode->setDisplayFrame(pSpriteFrame);

    w = pCCNode->getContentSize().width;
    h = pCCNode->getContentSize().height;

    if ( m_pAnimation->GetFrameCount() > 1)
    {
        m_pAnimateAction = new QAnimate();
        m_pAnimateAction->initWithAnimation( m_pAnimation->GetCCAnimation());
        m_pAnimateAction->SetTimeScale( timeScale);
        m_pRunAction = CCRepeatForever::create(m_pAnimateAction);
        m_pRunAction->retain();
        m_CCNode->runAction( m_pRunAction);
    }
}
//------------------------------------------------------------------------------
QAnimation *QSprite::get_animation(void)
{
    return m_pAnimation;
}
//------------------------------------------------------------------------------
void QSprite::set_frame(int frame)
{
    if ( m_pAnimateAction)
    {
        m_pAnimateAction->SetFrame( frame);

        // We don't want auto animation if we are manually setting the frame
        m_pAnimateAction->SetPlaying( false);

        // Update the width and height from the content
        CCSprite* pCCNode = (CCSprite*)m_CCNode;
        w = pCCNode->getContentSize().width;
        h = pCCNode->getContentSize().height;
    }
}
//------------------------------------------------------------------------------
int QSprite::get_frame(void)
{
    int frame = 1;

    if ( m_pAnimateAction)
    {
        frame = m_pAnimateAction->GetFrame();
    }

    return frame;
}
//------------------------------------------------------------------------------
void QSprite::GetBlendFunction( cocos2d::ccBlendFunc *blend) const
{
    if ( blendMode == "normal")
    {
        blend->src = GL_ONE; // assumes pre-multipled alpha
        blend->dst = GL_ONE_MINUS_SRC_ALPHA;
    }
    else if ( blendMode == "add")
    {
        blend->src = GL_ONE;
        blend->dst = GL_ONE;
    }
    else if ( blendMode == "multiply")
    {
        blend->src = GL_ZERO;
        blend->dst = GL_SRC_COLOR;
    }
    else if ( blendMode == "screen")
    {
        blend->src = GL_ONE;
        blend->dst = GL_ONE_MINUS_SRC_COLOR;
    }
    else
    {
        // Take from QAtlas blend mode
		if (m_pAnimation && m_pAnimateAction)
		{
			QAtlas* pAtlas = getAtlas();
			blend->src = pAtlas->_blendSrc;
			blend->dst = pAtlas->_blendDst;
		}
		else
		{
			// QSprite may just be a wrapper around CCSprite, with
			// no QAtlas etc. set up
			blend->src = ((CCSprite*)m_CCNode)->getBlendFunc().src;
			blend->dst = ((CCSprite*)m_CCNode)->getBlendFunc().dst;
		}
    }
}

QUICK_NAMESPACE_END;
